### PR TITLE
[fs/procfs]: Add procfs support  for error_report module in TizenRT

### PR DESCRIPTION
**Pull request by dr-venkman**
Add procfs module support for error_report module. Presently, TizenRT calls sched_gettcb function directly from application level. This commit replaces this through an appropriate procfs file read. In future, any attempt by error_report module to fetch parameters from call stack will go via the procfs implementation `ererport`.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [34f23b0bfab874607571aee7507a9b9349157402] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [34f23b0bfab874607571aee7507a9b9349157402] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [280da4986392b0ae33e51d07e206608ae911ec07] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [280da4986392b0ae33e51d07e206608ae911ec07] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4981a2767f588fb0015ef747f6a2d5885229be97] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4981a2767f588fb0015ef747f6a2d5885229be97] - Code Rule Check (C++) OK.

**Comment by Taejun-Kwon** [comment]
Our dev rule is that make driver for device.
This is kind of report, you have to expand procfs, instead of register virtual driver.

**Comment by seinfra** [comment]
Target : [cec4ba46f3cf0b401d7be62f05455580d7493cf8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [cec4ba46f3cf0b401d7be62f05455580d7493cf8] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [0f5a5821b563cf5ccc0037c751e418f37beed82a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [0f5a5821b563cf5ccc0037c751e418f37beed82a] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [2bd59ff35d2e3bc2005e90f5d1b9768667c74a73] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [2bd59ff35d2e3bc2005e90f5d1b9768667c74a73] - Code Rule Check Result:
os/fs/procfs/fs_procfsereport.c:194: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [b59da240ce06e39e70c2ed96bd7282557b575d6f] - Code Rule Check Result:
os/fs/procfs/fs_procfsereport.c:194: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [b59da240ce06e39e70c2ed96bd7282557b575d6f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e724c0cb2eeb9be3df458bfa23989f434f38949f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e724c0cb2eeb9be3df458bfa23989f434f38949f] - Code Rule Check Result:
os/fs/procfs/fs_procfsereport.c:194: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [8ce5043ac6770056e140606c9b033814700c792a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [8ce5043ac6770056e140606c9b033814700c792a] - Code Rule Check OK.

File: os/fs/procfs/fs_procfsereport.c [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+}
+
+/****************************************************************************
+ * Name: ereport_dup
+ *
+ * Description:
+ *   Duplicate open file data in the new file structure.
+ *
+ ****************************************************************************/
+static int ereport_dup(FAR const struct file *oldp, FAR struct file *newp)
```
Comment (Taejun-Kwon) [inline os/fs/procfs/fs_procfsereport.c]: You defined dup as a null, aren't you?

File: os/fs/procfs/fs_procfsereport.c [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+	if (index >= ereport_dir->base.nentries) {
+		/* We signal the end of the directory by returning the special
+		 * error -ENOENT
+		 */
+
+		fdbg("Entry %d: End of directory\n", index);
+		ret = -ENOENT;
+	}
+	/* No, we are not at the end of the directory */
+	else {
```
Comment (Taejun-Kwon) [inline os/fs/procfs/fs_procfsereport.c]: please apply coding rule.

File: external/error_report/error_report.c [hunk 3]
```diff
@@ -44,28 +50,33 @@ typedef struct {
...
 	int8_t q_pending;
+	bool is_mounted;
 	pthread_mutex_t err_mutex;
 	err_fsm_t fsm;
 } g_error_info_t;
 
 static g_error_info_t g_err_info;
 
-static unsigned long prv_fetch_taskaddr(int pid)
+err_status_t error_report_init(void)
```
[the remark this answers is not in the document]
Comment (dr-venkman) [reply external/error_report/error_report.c]: Thanks for the comment, will address the change in this commit.

File: external/error_report/error_report.c [hunk 4]
```diff
@@ -74,10 +85,36 @@ err_status_t error_report_init(void)
...
+#if !defined(CONFIG_FS_AUTOMOUNT_PROCFS)
+	if (!is_mounted) {
+		/* Detach mounted Procfs */
+		(void)umount(PROCFS_MOUNT_POINT);
+	}
+#endif
 	return ERR_SUCCESS;
 }
 
+static unsigned long prv_fetch_taskaddr(pid_t pid)
```
[the remark this answers is not in the document]
Comment (dr-venkman) [reply external/error_report/error_report.c]: You are right, `error_report` should go into framework. This migration will be handled in future as a separate pull request.

File: os/fs/procfs/fs_procfsereport.c (hunk 1 above)
Comment (dr-venkman) [reply os/fs/procfs/fs_procfsereport.c]: Yes, indeed. Will remove the implementation in this comment. Thanks for spotting.

File: os/fs/procfs/fs_procfsereport.c (hunk 2 above)
Comment (dr-venkman) [reply os/fs/procfs/fs_procfsereport.c]: Thanks, will update accordingly.

File: os/fs/procfs/Make.defs [hunk 5]
```diff
@@ -55,7 +55,7 @@ ifeq ($(CONFIG_FS_PROCFS),y)
 
 ASRCS +=
 CSRCS += fs_procfs.c fs_procfsutil.c fs_procfsproc.c fs_procfsuptime.c
-CSRCS += fs_procfscpuload.c fs_procfsversion.c
+CSRCS += fs_procfscpuload.c fs_procfsversion.c fs_procfsereport.c
```
[the remark this answers is not in the document]
Comment (dr-venkman) [reply os/fs/procfs/Make.defs]: Alright, I will add the ifeq for the following files: `fs_procfscpuload.c fs_procfsversion.c fs_procfsereport.c`.

File: os/fs/procfs/fs_procfsereport.c [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+	int i;
+
+	/* Two path forms are accepted:
+	 *
+	 * "ereport" - It is a top directory.
+	 * "ereport/<node>" - If <node> is a recognized node then, then it
+	 *   is a file or directory.
+	 */
+	fdbg("Inside function\n");
+	if (strncmp(relpath, "ereport", 7) != 0) {
```
[the remark this answers is not in the document]
Comment (dr-venkman) [reply os/fs/procfs/fs_procfsereport.c]: Thanks for the comment, will address the change in this commit.

File: os/fs/procfs/fs_procfs.c [hunk 7]
```diff
@@ -159,6 +160,11 @@ static const struct procfs_entry_s g_procfsentries[] = {
 #if defined(CONFIG_CM) && !defined(CONFIG_FS_PROCFS_EXCLUDE_CONNECTIVITY)
 	{"connectivity**", &cm_operations},
 #endif
+
+#if defined(CONFIG_ERROR_REPORT) && !defined(CONFIG_FS_PROCFS_EXCLUDE_EREPORT)
```
[the remark this answers is not in the document]
Comment (dr-venkman) [reply os/fs/procfs/fs_procfs.c]: I understand. If it is desired to have a meaningful text along with results, then indeed, error report module will have to parse (strstr calls) from the text. I can make output as `Error pid:<number>:task_addr:<address>`.

**Comment by seinfra** [comment]
Target : [d51463ca828d68bf6c197f0d5fbc7ae3e7585d97] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d51463ca828d68bf6c197f0d5fbc7ae3e7585d97] - Code Rule Check OK.

File: os/fs/procfs/fs_procfs.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (dr-venkman) [reply os/fs/procfs/fs_procfs.c]: Modified procfs as well as error_report module to generate the messages, and parse the data from them.

[the remark this answers is not in the document]
Comment (Taejun-Kwon) [reply os/fs/procfs/fs_procfs.c]: So now there's no dependency regarding CONFIG_ERROR_REPORT in procfs, you can remove it.

[the remark this answers is not in the document]
Comment (dr-venkman) [reply os/fs/procfs/fs_procfs.c]: We still need the check as we include `&ereport_operations` structure into `g_procfsentries`. Please correct me if I do not understand.

File: os/kernel/init/os_bringup.c [hunk 8]
```diff
@@ -286,7 +286,6 @@ static inline void os_do_appstart(void)
 		svdbg("Failed to create Task Manager\n");
 	}
 #endif
-
```
Comment (Taejun-Kwon) [inline os/kernel/init/os_bringup.c]: unnecessary changes.

File: os/fs/procfs/fs_procfs.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (dr-venkman) [reply os/fs/procfs/fs_procfs.c]: Modified the source and build as suggested. Thanks for clarifying.

File: os/kernel/init/os_bringup.c (hunk 8 above)
Comment (dr-venkman) [reply os/kernel/init/os_bringup.c]: Thanks for spotting. Removed the offending line.

**Comment by seinfra** [comment]
Target : [f98b2441a00737501f9911f49908a92400182f70] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f98b2441a00737501f9911f49908a92400182f70] - Code Rule Check OK.
